From 344fac80ace6b2691919a6751f06746ee74bef4a Mon Sep 17 00:00:00 2001
From: Stanislaw Gruszka <sgruszka@redhat.com>
Date: Wed, 10 Dec 2014 10:29:23 -0500
Subject: [wireless] brcmfmac: handle IF event for P2P_DEVICE interface

Message-id: <1418207389-7899-10-git-send-email-sgruszka@redhat.com>
Patchwork-id: 101581
O-Subject: [RHEL7.1 PATCH 09/35] brcmfmac: handle IF event for P2P_DEVICE interface
Bugzilla: 1155538
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: John Greene <jogreene@redhat.com>

From: Arend van Spriel <arend@broadcom.com>

BZ#1155538

commit 87c4790330810fe5caf0172d9320cf24ef19cebe upstream.

The firmware notifies about interface changes through the IF event
which has a NO_IF flag that means host can ignore the event. This
behaviour was introduced in the driver by:

  commit 2ee8382fc6c763c76396a6aaff77a27089eed3aa
  Author: Arend van Spriel <arend@broadcom.com>
  Date:   Sat Aug 10 12:27:24 2013 +0200

      brcmfmac: ignore IF event if firmware indicates it

It turns out that the IF event for the P2P_DEVICE also has this
flag set, but the event should not be ignored in this scenario.
The mentioned commit caused a regression in 3.12 kernel in creation
of the P2P_DEVICE interface.

Reviewed-by: Hante Meuleman <meuleman@broadcom.com>
Reviewed-by: Franky (Zhenhui) Lin <frankyl@broadcom.com>
Reviewed-by: Daniel (Deognyoun) Kim <dekim@broadcom.com>
Reviewed-by: Pieter-Paul Giesberts <pieterpg@broadcom.com>
Signed-off-by: Arend van Spriel <arend@broadcom.com>
Signed-off-by: John W. Linville <linville@tuxdriver.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Brew:
https://brewweb.devel.redhat.com/taskinfo?taskID=8330268

Testing:
Compiled only.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/wireless/brcm80211/brcmfmac/fweh.c b/drivers/net/wireless/brcm80211/brcmfmac/fweh.c
index fad77dd..3f9cb89 100644
--- a/drivers/net/wireless/brcm80211/brcmfmac/fweh.c
+++ b/drivers/net/wireless/brcm80211/brcmfmac/fweh.c
@@ -185,7 +185,13 @@ static void brcmf_fweh_handle_if_event(struct brcmf_pub *drvr,
     ifevent->action, ifevent->ifidx, ifevent->bssidx,
     ifevent->flags, ifevent->role);
 
- if (ifevent->flags & BRCMF_E_IF_FLAG_NOIF) {
+ /* The P2P Device interface event must not be ignored
+  * contrary to what firmware tells us. The only way to
+  * distinguish the P2P Device is by looking at the ifidx
+  * and bssidx received.
+  */
+ if (!(ifevent->ifidx == 0 && ifevent->bssidx == 1) &&
+     (ifevent->flags & BRCMF_E_IF_FLAG_NOIF)) {
   brcmf_dbg(EVENT, "event can be ignored\n");
   return;
  }
@@ -210,12 +216,12 @@ static void brcmf_fweh_handle_if_event(struct brcmf_pub *drvr,
     return;
  }
 
- if (ifevent->action == BRCMF_E_IF_CHANGE)
+ if (ifp && ifevent->action == BRCMF_E_IF_CHANGE)
   brcmf_fws_reset_interface(ifp);
 
  err = brcmf_fweh_call_event_handler(ifp, emsg->event_code, emsg, data);
 
- if (ifevent->action == BRCMF_E_IF_DEL) {
+ if (ifp && ifevent->action == BRCMF_E_IF_DEL) {
   brcmf_fws_del_interface(ifp);
   brcmf_del_if(drvr, ifevent->bssidx);
  }
diff --git a/drivers/net/wireless/brcm80211/brcmfmac/fweh.h b/drivers/net/wireless/brcm80211/brcmfmac/fweh.h
index 51b53a7..d26b476 100644
--- a/drivers/net/wireless/brcm80211/brcmfmac/fweh.h
+++ b/drivers/net/wireless/brcm80211/brcmfmac/fweh.h
@@ -167,6 +167,8 @@ enum brcmf_fweh_event_code {
 #define BRCMF_E_IF_ROLE_STA   0
 #define BRCMF_E_IF_ROLE_AP   1
 #define BRCMF_E_IF_ROLE_WDS   2
+#define BRCMF_E_IF_ROLE_P2P_GO   3
+#define BRCMF_E_IF_ROLE_P2P_CLIENT  4
 
 /**
  * definitions for event packet validation.
-- 
1.7.1